pkg/retry/retry.go

```diff
@@ -34,12 +36,13 @@ type Settings struct {
 func WithBackoff(
 	ctx context.Context, retryableFunc RetryableFunc, retryable IsRetryable, b backoff.Backoff, settings Settings,
 ) (err error) {
-	parentCtx := ctx
+	timeWindow, cancelTimeWindow := context.WithCancel(context.Background())
+	defer cancelTimeWindow()
 
 	if settings.Timeout > 0 {
-		var cancelCtx context.CancelFunc
-		ctx, cancelCtx = context.WithTimeout(ctx, settings.Timeout)
-		defer cancelCtx()
+		defer time.AfterFunc(settings.Timeout, func() {
+			cancelTimeWindow()
+		}).Stop()
 	}
```

Comment:
`timeWindow` isn't really used like a context but more like a `time.Timer`. I think I'd change it to be that because otherwise, there is the question why it derives from `context.Background()` instead of `ctx`. Probably because otherwise the logging would have to be adapted to distinguish between both cancellation reasons. Using `time.Timer` instead would simply avoid the confusion.

```go
var timeout <-chan Time // receiving from nil channel blocks forever, acting like no timeout
if settings.Timeout > 0 {
	t := time.NewTimer(settings.Timeout)
	defer t.Stop()
	timeout = t.C
}

// ...
select {
// ...
case <-timeout:
```

Reply:
Done.